달력 조회 서비스 작성

jiwon83 opened the pull request:
## 📑 개요

#38 

와플 v1을 개선하며 바뀐 반복되는 일정에 대한 로직의 변화에 따라 달력 조회 기능을 새롭게 구현했습니다.

### AS-IS
- v1에서는 반복되는 일정을 반복 갯수만큼 일정 데이터로 저장하여 관리했기 때문에 조회 로직은 매우 단순했습니다.
- 기존 반복 저장 로직에서도 다소 복잡한 분기문이 존재했습니다. [v1관련코드](https://github.com/What-s-Your-Plan/wypl-lab/blob/main/backend/src/main/java/com/butter/wypl/schedule/service/ScheduleServiceImpl.java) _line 174 ~ 266_

### TO-BE
- 반복 일정을 하나의 Schedule 객체로 관리하도록 변경함으로써 저장에서 발생하는 서버 부하를 줄였습니다. 
- 이에 따라, 이전보다 조회 로직이 다소 복잡해졌습니다. 따라서 하나의 서비스에서 모든 로직을 구현하지 않고 전략 패턴을 적용하여 코드를 분리, 확장성과 가독성을 향상시키는 방안을 고려했습니다.
<br>

다음 기준에 따라서 전략 패턴을 적용했습니다.
```java
 - 달력별 다른 조회 기간 설정
 - 반복주기에 따라서 조회 조건에 맞게 가공
```


### ✅ PR 체크리스트

---

- [x] 🔀 PR 제목의 형식을 잘 작성했나요?
- [x] 💯 테스트는 잘 통과했나요?
- [x] 🏗️ 빌드는 성공했나요?
- [x] 🧹 불필요한 코드는 제거했나요?
- [x] 💭 이슈는 등록했나요?
- [x] 🏷️ 라벨은 등록했나요?

### 🚀 상세 작업 내용

---

<!-- 상세 작업 내용 에서는 어떠한 작업을 했는지 상세히 적어주세요! -->

<!-- 예시는 다음과 같습니다. -->
<!-- - [X] 아이디 중복 검사 비즈니스 로직 구현 -->
<!-- - [X] 중복일 경우 예외 처리 기능 구현 -->

### 📁 ETC

---

<!-- 이 곳에서는 관련 자료나 사진을 올여주세요! -->
<!-- 링크를 넣고 싶은 경우에는 MAC 에서는 커맨드 + K, Windows 에서는 컨트롤 + K를 누르면 [](url) 가 생성되는데 [] 안에는 원하시는 링크의 제목을 입력하고 () 안에는 URL을
입력해주세요! -->
<!-- 사진 같은 경우에는 drag and drop 으로 사진을 추가할 수 있습니다! -->

jiwon83 committed:
feat: 캘린더 조회 전략 패턴 적용

Resolves: #38

jiwon83 committed:
Merge remote-tracking branch 'origin/dev' into feat/#38

jiwon83 committed:
chore: querydsl 의존성 추가

Resolves: #38

jiwon83 committed:
feat: ScheduleRepository 에서 calendarId, startDate, endDate 로 List<Schedule> 조회 기능 추가

Resolves: #38

jiwon83 committed:
feat: CalendarService 추가

- 전략 + 플라이웨이트 패턴 적용 테스트 포함

Resolves: #38

jiwon83 committed:
feat: CalendarServiceUtil의 Day, Week 반복 일정 조회 기능 구현

- CalendarServiceUtil.java
- CalendarServiceTest.java

Resolves: #38

jiwon83 committed:
Test: Calendar, Schedule, Scheduleinfo Fixture 생성

Resolves: #38

jiwon83 committed:
chore: wypl-core 모듈 lombok 설정

Resolves: #38

jiwon83 committed:
feat: 반복 주기 = Month 조회 기능 구현 in CalendarServiceUtil

- CalendarServiceUtil.java
- CalendarServiceUtilTest.java

Resolves: #38

jiwon83 committed:
fix: 반복 주기 = Week 조회 기능의 버그 수정

- 검색 조건이 반복 주에 속하지 않은 케이스에 대한 버그 수정을 위해 getNearestDateUsingWeekInterval 메서드 추가 구현

Resolves: #38

jiwon83 committed:
feat: 반복 주기 = Year 조회 기능 구현 in CalendarServiceUtil

Resolves: #38

jiwon83 committed:
refactor: 반복주기에 따른 할일 조회 기능 리팩토링

- 기존의 Util Class의 메서드로 존재하던 구조 대신 전략 패턴 적용해서 클래스 단위로 분리
- 이에 따른 테스트 수정

Resolves: #38

jiwon83 committed:
feat: Calendar Type에 따른 전략 클래스 구현

Resolves: #38

jiwon83 committed:
refactor: Calendar Type에 따른 전략 클래스 구현

- CalendarStrategy의 ScheduleRepository 연관관계 제거
- 주석 추가

Resolves: #38

jiwon83 committed:
feat: ScheduleFindResponse.java 구현

Resolves: #38

jiwon83 committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into feat/#38

KIMSEI1124 commented on application/wypl-core/build.gradle:
```diff
@@ -17,4 +17,9 @@ dependencies {
+    compileOnly 'org.projectlombok:lombok:1.18.24'
+    annotationProcessor 'org.projectlombok:lombok:1.18.24'
+    testCompileOnly 'org.projectlombok:lombok:1.18.24'
+    testAnnotationProcessor 'org.projectlombok:lombok:1.18.24'
```
`Lombok`은 기본으로 적용이 되어있습니다!

해당 내용은 없어도 될 것 같아요

jiwon83 replied:
testAnnotationProcessor 만 추가하도록 변경했습니다.

jiwon83 committed:
fix: fix build error

Resolves: #38

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/calendar/CalendarServiceUtil.java:
```diff
@@ -0,0 +1,192 @@
+public class CalendarServiceUtil {
+
+	public static List<ScheduleFindResponse>  getScheduleResponses(Schedule schedule, LocalDate startDate, LocalDate endDate) {
+		if (!schedule.isRepetition()) return List.of(ScheduleFindResponse.of(schedule, schedule.getStartDateTime(), schedule.getEndDateTime()));
```
```suggestion
		if (!schedule.isRepetition()) {
			return List.of(ScheduleFindResponse.of(schedule, schedule.getStartDateTime(), schedule.getEndDateTime()));
		}
```

이렇게 단 한줄이라도 if문이 있다면 다음과 같이 해주시면 좋을 것 같아요!

jiwon83 replied:
그렇게 수정하겠습니다 감사합니다!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/calendar/CalendarServiceUtil.java:
```diff
@@ -0,0 +1,192 @@
+			LocalDateTime startDateTime = LocalDateTime.of(searchStartDate, schedule.getStartDateTime().toLocalTime());
+			LocalDateTime endDateTime = LocalDateTime.of(searchStartDate, schedule.getEndDateTime().toLocalTime());
+			responses.add(ScheduleFindResponse.of(schedule, startDateTime, endDateTime));
+		}
```
해당 코드도 좋지만 추후 여유가 되신다면 `stream`을 한번 사용해보는게 좋을 것 같습니다!

가독성이 훨씬 좋아질 것 같아요!

jiwon83 replied:
리뷰 감사합니다. 적용해 보겠습니다 : )

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/calendar/CalendarServiceUtil.java:
```diff
@@ -0,0 +1,192 @@
+			if( isSelectedDayOfWeek(repetitionDayOfWeek, dayOfWeek) ) {
+
+				// Find nearest day by dayOfWeek and increase weekInterval
+				LocalDate date = getMaxDate(searchStartDate, schedule.getRepetitionStartDate()).with(TemporalAdjusters.nextOrSame(DayOfWeek.of(dayOfWeek)));
```
개인적으로 `TemporalAdjusters.nextOrSame(DayOfWeek.of(dayOfWeek))` 는 직관적이지 않은 것 같습니다!

혹시 메서드로 분리하는건 어떻게 생각하시나요?

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/calendar/service/CalendarService.java:
```diff
@@ -0,0 +1,53 @@
+     * @param calendarFindRequest
+     * @return FindCalendarResponse
+     */
+    @Transactional
```
제가 보기에는 단순 조회문인데 트랜잭션이 `readOnly`가 아닌 이유가 있을까요?

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/calendar/service/strategy/DayCalendarStrategy.java:
```diff
@@ -0,0 +1,45 @@
+
+        for (Schedule schedule : schedules) {
+            RepetitionService.getScheduleResponses(schedule, startDate, startDate);
+        }
```
이런 것도 추후 `stream`으로 간단하게 처리할 수 있는데 추후 한번 도전해보시면 좋을 것 같아요!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/RepetitionStrategyFactory.java:
```diff
@@ -0,0 +1,37 @@
+	 * @param repetitionCycle
+	 * @return RepetitionStrategy
+	 */
+	public static RepetitionStrategy getRepetitionStrategy(RepetitionCycle repetitionCycle) {
+		if (instance == null) {
+			instance = new RepetitionStrategyFactory();
+		}
+		return instance.map.get(repetitionCycle);
+	}
```
싱글톤패턴 잘 사용하신 것 같습니다! 👍🏼

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/strategy/MonthRepetitionStrategy.java:
```diff
@@ -0,0 +1,46 @@
+		searchEndDate = getMinDate(searchEndDate, schedule.getRepetitionEndDate());
+
+		// 끝나는 날짜
+		int endDayOfMonth = schedule.getEndDateTime().getDayOfMonth();
```
음.. 해당 내용에서 캡슐화가 필요한 것 같습니다!

schedule에서 endDateTime 을 꺼내고 거기서 month를 가져오는데 이를 schedule에서 메서드로 만든 후 `getEndMonth()`와 같은 함수로 호출하는게 더 좋을 것 같아요!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/strategy/MonthRepetitionStrategy.java:
```diff
@@ -0,0 +1,46 @@
+		int endDayOfMonth = schedule.getEndDateTime().getDayOfMonth();
+
+		// 탐색할 시작 일시와 끝 일시 설정
+		LocalDateTime searchEndDateTime = LocalDateTime.of(searchStartDate.withDayOfMonth(endDayOfMonth), schedule.getEndDateTime().toLocalTime());
```
이것도 동일하게 캡슐화를 진행하면 좋을 것 같습니다!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/strategy/MonthRepetitionStrategy.java:
```diff
@@ -0,0 +1,46 @@
+		LocalDateTime searchStartDateTime = searchEndDateTime.minus(duration);
+
+		for ( LocalDate date = searchStartDateTime.toLocalDate(); !date.isAfter(searchEndDate); date = date.plusMonths(1)) {
+			LocalDateTime startDateTime = LocalDateTime.of(date, schedule.getStartDateTime().toLocalTime());
```
이것도 동일하게 캡슐화를 진행하면 좋을 것 같습니다! 🥈

KIMSEI1124 commented on domain/jpa-common/build.gradle:
```diff
@@ -10,4 +10,31 @@ dependencies {
 
     runtimeOnly('com.h2database:h2')
+
+    /* MapStruct */
+    api('org.mapstruct:mapstruct:1.5.5.Final')
+    annotationProcessor 'org.mapstruct:mapstruct-processor:1.5.5.Final'
```
혹시 해당 라이브러리를 사용하고 계신가요?! 아니라면 삭제해주시면 감사합니다!

KIMSEI1124 commented on domain/jpa-calendar-domain/src/main/java/com/wypl/jpacalendardomain/calendar/repository/ScheduleRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,27 @@
+public class ScheduleRepositoryCustomImpl implements ScheduleRepositoryCustom {
+
+	private final JPAQueryFactory jpaQueryFactory;
+	private final QSchedule schedule = QSchedule.schedule;
```
해당 내용을 `import static` 하는 방법이 있는데 해당 방법을 한번 보시면 좋을 것 같아요!

jiwon83 replied:
적용했습니다!

KIMSEI1124 commented on domain/jpa-calendar-domain/src/main/java/com/wypl/jpacalendardomain/calendar/repository/ScheduleRepository.java:
```diff
@@ -3,5 +3,5 @@
 import org.springframework.data.jpa.repository.JpaRepository;
 
-public interface ScheduleRepository extends JpaRepository<Schedule, Long> {
+public interface ScheduleRepository extends JpaRepository<Schedule, Long>, ScheduleRepositoryCustom {
```
QueryDSL 좋습니다! 👍🏼 

추후 해당 방법이 아닌 다른 방법도 있는데 한번 알아보시면 좋을 것 같아요~ 이러한 케이스에서는 괜찮지만 복잡한 케이스가 된다면 다른 방법이 더 효율적일 수 있습니다!

KIMSEI1124 commented on application/wypl-core/src/test/java/com/wypl/wyplcore/calendar/service/strategy/CalendarStrategyConfigTest.java:
```diff
@@ -0,0 +1,45 @@
+        for (Map.Entry<?, ?> entry : instance.entrySet()) {
+            assertThat(entry.getKey()).isInstanceOf(CalendarType.class);
+            assertThat(entry.getValue()).isInstanceOf(CalendarStrategy.class);
+            System.out.println(entry.getKey() + " : " + entry.getValue());
```
시스템을 호출하여 출력하는 것 보다 slf4j를 활용하여 `log`를 남기는게 더 좋을 것 같습니다!

jiwon83 committed:
Refactor(#38): refactor code related to calendar strategy

Resolves: #38

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/strategy/WeekRepetitionStrategy.java:
```diff
@@ -0,0 +1,122 @@
+				LocalDate date = getMaxDate(searchStartDate, schedule.getRepetitionStartDate()).with(
+					TemporalAdjusters.nextOrSame(DayOfWeek.of(dayOfWeek)));
+
+				for(; !date.isAfter(searchEndDate); date = date.plusWeeks(schedule.getWeekInterval())) {
```
`!date.isAfter(searchEndDate)` 이런 조건식을 메서드로 분리해보면 어떨까요?

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/strategy/WeekRepetitionStrategy.java:
```diff
@@ -0,0 +1,122 @@
+
+		// searchStartDate를 해당 주의 월요일로 설정
+		LocalDateTime searchStartMonday = LocalDateTime.of(searchStartDate.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY)), LocalTime.of(0, 0));
+		LocalDateTime scheduleStartMonday = LocalDateTime.of(schedule.getRepetitionStartDate().with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY)), LocalTime.of(0, 0));
```
`schedule.getRepetitionStartDate().with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY)` 해당 함수도 메서드로 분리하면 어떨까요?!

매개변수로 두 개만 넘기는 형식으로 하면 좋을 것 같아요.

KIMSEI1124 reviewed (changes requested):
개발하시느라 고생많으셨습니다! 😄 
전체적인 코드는 잘 작성하였습니다! 👍🏼 

하지만 컨벤션 및 함수로 분리하면 좋을 것 같은 내용들이 있었습니다.
해당 내용들은 자세히 작성해두었고, 더 필요한 부분은 노션에다가 따로 작성해놨습니다!

고생 많으셨고 해당 내용 수정해주시면 감사합니다!

jiwon83 committed:
Merge remote-tracking branch 'origin/dev' into feat/#38

# Conflicts:
#	application/wypl-core/build.gradle

jiwon83 committed:
refactor(#38): refactor & remove unnecessary code

Resolves: #38

jiwon83 committed:
chore(#38): add dependency "testAnnotationProcessor"

Resolves: #38

jiwon83 committed:
feat(#38): add CalendarRepository

Resolves: #38

jiwon83 committed:
refactor(#38): reformat code for applying code convention

Resolves: #38

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/calendar/service/CalendarService.java:
```diff
@@ -0,0 +1,86 @@
+
+	private final ScheduleRepository scheduleRepository;
+
+	private final Map<CalendarType, CalendarStrategy> calendarStrategyMap;
```
혹시 개인적인 의견이지만 `calendarStrategies`로 수정하는게 어떨까요?! 변수명에 자료구조 형식이 드러나는 것이 좋아보이지 않습니다!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/calendar/service/CalendarService.java:
```diff
@@ -0,0 +1,86 @@
+	 * @return FindCalendarResponse
+	 */
+	@Transactional
+	public CalendarSchedulesResponse findCalendar(AuthMember authMember, long calendarId,
+		CalendarFindRequest calendarFindRequest) {
```
```suggestion
	public CalendarSchedulesResponse findCalendar(
		AuthMember authMember,
		long calendarId,
		CalendarFindRequest calendarFindRequest
	) {
```

이런 형식으로 매개변수를 작성해보는건 어떨까요?!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/calendar/service/CalendarServiceUtil.java:
```diff
@@ -0,0 +1,15 @@
+
+public class CalendarServiceUtil {
+
+	public static LocalDate getMaxDate(LocalDate date1, LocalDate date2) {
```
음 혹시 `getNewestDate`는 어떨까요?? 아래 내용도 `getOldestDate`도 괜찮을 것 같다고 생각합니다!

해당 반환값이 `Number`형식이 아닌 날짜 형식이기 때문에 메서드명을 수정하시는게 어떨까 생각합니다!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/calendar/service/strategy/CalendarStrategyConfig.java:
```diff
@@ -0,0 +1,29 @@
+		});
+	}
+
+	@Bean
+	public Map<CalendarType, CalendarStrategy> calendarStrategyMap() {
+		return calendarStrategyMap;
+	}
```
`CalendarService`에서 피드백을 준 내용을 토대로 수정한다면 해당 `@Bean`의 이름이 수정되어야 합니다!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/controller/ScheduleController.java:
```diff
@@ -19,11 +19,12 @@
-        return WyplResponseEntity.created(response, "일정이 생성됐습니다.");
-    }
+	@PostMapping
+	public WyplResponseEntity<ScheduleInfoCreateResponse> addSchedule(@Authenticated AuthMember authMember,
+		@RequestBody ScheduleCreateRequest scheduleCreateRequest) {
```
해당 내용도 위에는 있지만!

```suggestion
	public WyplResponseEntity<ScheduleInfoCreateResponse> addSchedule(
		@Authenticated AuthMember authMember,
		@RequestBody ScheduleCreateRequest scheduleCreateRequest
	) {
```
와 같이 수정하는게 더 보기 편하다고 생각하는데 어떠신가요?

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/data/response/ScheduleInfoCreateResponse.java:
```diff
@@ -2,13 +2,13 @@
-        @JsonProperty("calendar_id")
-        Long calendarId,
+	@JsonProperty("calendar_id")
+	Long calendarId,
```
혹시 `long`이 아닌 `Long`으로 사용하신 이유가 있을까요?!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/RepetitionService.java:
```diff
@@ -0,0 +1,25 @@
+	 */
+	public static List<ScheduleFindResponse> getScheduleResponses(Schedule schedule, LocalDate searchStartDate,
+		LocalDate searchEndDate) {
+		RepetitionStrategy repetitionStrategy = RepetitionStrategyFactory.getRepetitionStrategy(
+			schedule.getRepetitionCycle());
```
해당 내용을 보면서 느낀점이 이미 생성되어 있는 객체를 가져오고 있는데 팩토리라는 이름은 이상한 것 같습니다.
팩토리패턴은 새로운 객체를 생성할 때 사용해야한다고 생각하는데 해당 패턴은 [Service locator pattern](https://www.google.com/search?q=%EC%84%9C%EB%B9%84%EC%8A%A4+%EB%A1%9C%EC%BC%80%EC%9D%B4%ED%84%B0+%ED%8C%A8%ED%84%B4&oq=%EC%84%9C%EB%B9%84%EC%8A%A4+%EB%A1%9C%EC%BC%80%EC%9D%B4%ED%84%B0+%ED%8C%A8%ED%84%B4&gs_lcrp=EgZjaHJvbWUyBggAEEUYOTIKCAEQABiABBiiBDIKCAIQABiABBiiBDIKCAMQABiABBiiBNIBBzIzOWowajGoAgCwAgA&sourceid=chrome&ie=UTF-8)을 참고해주시면 감사합니다!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/strategy/DayRepetitionStrategy.java:
```diff
@@ -0,0 +1,38 @@
+		searchStartDate = getMaxDate(searchStartDate, schedule.getRepetitionStartDate());
+		searchEndDate = getMinDate(searchEndDate, schedule.getRepetitionEndDate());
+
+		for (; !searchStartDate.isAfter(searchEndDate); searchStartDate = searchStartDate.plusDays(1)) {
```
`!searchStartDate.isAfter(searchEndDate)`여기서 의문이 들었던 점이 있습니다!
`isBefore`로 처리하고 `!`를 없앨 수 있을 것 같은데 이렇게 하신 이유가 있을까요?

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/strategy/WeekRepetitionStrategy.java:
```diff
@@ -0,0 +1,128 @@
+
+		int gapOfWeek = (int)Duration.between(scheduleStartMonday, searchStartMonday).toDays() / 7;
+
+		if (gapOfWeek % weekInterval == 0)
+			return searchStartDate;
```
단 한줄이여도 괄호 부탁드립니다. 🙏

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/strategy/WeekRepetitionStrategy.java:
```diff
@@ -0,0 +1,128 @@
+				LocalDate date = getMaxDate(searchStartDate, schedule.getRepetitionStartDate()).with(
+					TemporalAdjusters.nextOrSame(DayOfWeek.of(dayOfWeek)));
+
+				for (; !date.isAfter(searchEndDate); date = date.plusWeeks(schedule.getWeekInterval())) {
```
해당 조건절도 `isBefore`로 처리가 될 것 같은데 한번만 확인 부탁드립니다!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/schedule/service/repetition/strategy/YearRepetitionStrategy.java:
```diff
@@ -0,0 +1,62 @@
+	 * @return LocalDateTime
+	 */
+	private static LocalDateTime getNearestStartDateTime(Schedule schedule, LocalDate searchStartDate) {
+		LocalDateTime nearestEndDateTime = LocalDateTime.of(
+			searchStartDate.withDayOfYear(schedule.getEndDateTime().getDayOfYear()),
+			schedule.getEndDateTime().toLocalTime());
```
`schedule.getEndDateTime().getDayOfYear()`를 캡슐화해서 메서드 체이닝이 단 한번만 호출해서 데이터를 가져오도록 캡슐화를 진행하면 좋을 것 같습니다!

KIMSEI1124 commented on application/wypl-core/src/test/java/com/wypl/wyplcore/ScheduleFixture.java:
```diff
@@ -0,0 +1,135 @@
+			.build();
+	}
+
+	public Schedule toObject() {
+		return Schedule.builder()
+			.title(this.title)
+			.description(this.description)
+			.startDateTime(this.startDateTime)
+			.endDateTime(this.endDateTime)
+			.repetitionStartDate(this.repetitionStartDate)
+			.repetitionEndDate(this.repetitionEndDate)
+			.repetitionCycle(this.repetitionCycle)
+			.dayOfWeek(this.dayOfWeek)
+			.weekInterval(this.weekInterval)
+			.build();
+	}
```
`toObject`인데 반환값은 `Schedule`입니다! 어떠한 역할을 하는 메서드일까요? 궁금합니다!

KIMSEI1124 commented on build.gradle:
```diff
@@ -34,6 +34,7 @@ subprojects {
         compileOnly 'org.projectlombok:lombok'
         testImplementation 'org.projectlombok:lombok'
         annotationProcessor 'org.projectlombok:lombok'
+        testAnnotationProcessor 'org.projectlombok:lombok'
```
대신 넣어주셔서 감사합니다! 💯

KIMSEI1124 reviewed (changes requested):
몇몇 코멘트를 달아뒀습니다!
한 번 확인하시고 궁금하신 점 있으면 코멘트 부탁드립니다!

jiwon83 committed:
feat(#38): 달 반복 일정 조회 -

- findNextDayOfMonth 추가
- MonthRepetitionStrategyTest의 getSchedulesResponsesForSpecial1, getSchedulesResponsesForSpecial2

Resolves: #38

jiwon83 committed:
refactor(#38): Schedule 엔티티의 getDuration method 추가

Resolves: #38

jiwon83 committed:
fix(#38): Week 반복주기 버그 개선

Resolves: #38

jiwon83 committed:
refactor(#38): Stream 적용 및 리팩토링

Resolves: #38

jiwon83 committed:
refactor(#38): Q클래스 static import 적용

Resolves: #38

jiwon83 committed:
feat(#38): dayOfYear과 dayOfWeek를 매개변수로 갖는 getNextOrSame 기능 구현

Resolves: #38

jiwon83 committed:
feat(#38): Add DateUtil.java

Resolves: #38

jiwon83 committed:
refactor(#38): Year 반복 조회를 위한 findNextOrSame 수정

as-is : dayOfYear 사용 -> Month, dayOfMonth 로 직접 구현

Resolves: #38

jiwon83 committed:
refactor(#38): RepetitionStrategy 리팩토링

Resolves: #38

KIMSEI1124 reviewed (approved):
LGTM! 👍🏼

KIMSEI1124 committed:
Merge branch 'dev' into feat/#38

KIMSEI1124 committed:
Fix(#38): 테스트 오류 해결

- Setter 엑세스 레벨 수정
- ymal 파일 수정
- Boot Test 설정 수정

KIMSEI1124 committed:
Chore(#38): github actions 수정 필요
